### PR TITLE
98200 LTC json to xml mapping migrate new licence

### PR DESCRIPTION
**Pull request by sakshi15525**
1. Changed attribute 'upn' to 'ecdisManufacturerName' for both licenceupdated and ros events.
2. Json to xml mapping for migrate new licence events.
3. Added logging.
4. Added Unit test cases.
5. Added functional tests

### CODE REVIEW

File: src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs [hunk 1]
```diff
@@ -17,6 +17,7 @@ public class RecordOfSaleSapMessageBuilder : IRecordOfSaleSapMessageBuilder
         private const string ImOrderNameSpace = "RecordOfSale";
         private const string MaintainHoldingsType = "MAINTAINHOLDINGS";
         private const string NewLicenceType = "NEWLICENCE";
+        private const string MigrateNewLicenseType = "MIGRATENEWLICENCE";
```
Comment (pravinvarade10801) [inline src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs:20]:
```suggestion
        private const string MigrateNewLicenceType = "MIGRATENEWLICENCE";
```

File: src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs [hunk 2]
```diff
@@ -48,6 +49,7 @@ public XmlDocument BuildRecordOfSaleSapMessageXml(List<RecordOfSaleEventPayLoad>
             {
                 NewLicenceType => BuildNewLicencePayload(eventDataList),
                 MaintainHoldingsType => BuildMaintainHoldingsPayload(eventDataList),
+                MigrateNewLicenseType => BuildMigrateNewLicencePayload(eventDataList),
```
Comment (pravinvarade10801) [inline src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs:52]:
```suggestion
                MigrateNewLicenceType => BuildMigrateNewLicencePayload(eventDataList),
```

File: src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs [hunk 3]
```diff
@@ -182,5 +184,66 @@ private SapRecordOfSalePayLoad BuildMaintainHoldingsPayload(List<RecordOfSaleEve
...
+                        Duration = rosUnitOfSale.Duration,
+                        ReNew = rosUnitOfSale.ReNew,
+                        Repeat = string.Empty
+                    })
+                        .ToList();
+
+                    prod.UnitOfSales = unitOfSaleList;
+                    rosMigrateNewLicencePayload.PROD = prod;
+                }
+
```
Comment (pravinvarade10801) [inline src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs:231]: Remove this extra line

**Review by pravinvarade10801 — COMMENTED** [review]
Have few minor comments.

File: src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs (hunk 1 above)
Comment (sakshi15525) [reply src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs:20]: Done

File: src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs (hunk 2 above)
Comment (sakshi15525) [reply src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs:52]: done

**Review by pravinvarade10801 — APPROVED** [review]
LGTM

File: tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/RoSWebhookScenarios.cs [hunk 4]
```diff
@@ -73,6 +73,7 @@ public async Task WhenValidRoSEventInRecordOfSalePublishedEventPostReceivedWithI
         }
 
         [Test, Order(2)]
+        [TestCase("LTC03_InValidLTCwithNoCorrelationID.json", TestName = "WhenValidRoSEventPublishedEventReceivedInLTCMigrateNewLicencePostReceivedWithInvalidPayload_ThenWebhookReturns400BadRequestResponse")]
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/RoSWebhookScenarios.cs:76]:
```suggestion
        [TestCase("LTC03_InValidLTCwithNoCorrelationID.json", TestName = "WhenValidRoSMigrateNewLicencePublishedEventReceivedWithInvalidPayload_ThenWebhookReturns400BadRequestResponse")]
```

File: tests/UKHO.ERPFacade.EventAggregation.WebJob.Tests/Helpers/RecordOfSaleSapMessageBuilderTests.cs [hunk 5]
```diff
@@ -84,12 +84,17 @@ public class RecordOfSaleSapMessageBuilderTests
...
...ForMaintainHoldings1 = @"{""specversion"":""1.0"",""type"":""uk.gov.ukho.shop.recordOfSale.v1"",""source"":""https://uk.gov.ukho.shop"",""id"":""e744fa37-0c9f-4795-adc9-7f42ad8f11c1"",""time"":""2023-07-20T10:40:00Z"",""subject"":""releasability set changes holdings Record of Sale"",""datacontenttype"":""application/json"",""data"":{""correlationId"":""123-abc-456-xyz-333"",""relatedEvents"":[""e744fa37-0c9f-4795-adc9-7f42ad8f11c1"", ""e744fa37-0c9f-4795-adc9-7f42ad8f1234""],""recordsOfSale"":{""licenseId"":"""",""productType"":""AVCS"",""transactionType"":""MAINTAINHOLDINGS"",""distributorCustomerNumber"":"""",""shippingCoNumber"":"""",""ordernumber"":""5432796"",""orderDate"":"""",""po-ref"":""75277T"",""holdingsExpiryDate"":"""",""sapId"":""75277T"",""vesselName"":"""",""imoNumber"":"""",""callSign"":"""",""licenceType"":"""",""shoreBased"":"""",""fleetName"":"""",""numberLicenceUsers"":null,""ecdisManufacturerName"":"""",""licenceDuration"":null,""unitsOfSale"":[{""unitName"":""PT111101"",""endDate"":""2023-10-31"",""duration"":""3"",""renew"":""E"",""repeat"":""P""},{""unitName"":""GB302409"",""endDate"":""2023-12-01"",""duration"":""6"",""renew"":""E"",""repeat"":""P""}]}}}";
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.EventAggregation.WebJob.Tests/Helpers/RecordOfSaleSapMessageBuilderTests.cs:91]: Can we give meaningful name to variables based on scenario.

File: tests/UKHO.ERPFacade.EventAggregation.WebJob.Tests/Helpers/RecordOfSaleSapMessageBuilderTests.cs [hunk 6]
```diff
@@ -313,6 +318,87 @@ public void WhenTransactionTypeIsMaintainHoldings_ThenReturns_SomeFieldsEmptyInS
...
             result.LicenceType.Should().Be("");
             result.PROD.UnitOfSales.Count.Should().Be(4);
+            result.PROD.UnitOfSales[0].Repeat.Should().Be("P");
+            result.PROD.UnitOfSales[2].Repeat.Should().Be("P");
+        }
+
+        [Test]
+        public void WhenTransactionTypeIsMigrateNewLicence_ThenReturnXMLDocument()
+        {
+            var jsonData1 = JsonConvert.DeserializeObject<RecordOfSaleEventPayLoad>(jsonForMigrateNewLicence1);
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.EventAggregation.WebJob.Tests/Helpers/RecordOfSaleSapMessageBuilderTests.cs:328]: Meaningful naming conventions here as well

File: tests/UKHO.ERPFacade.EventAggregation.WebJob.Tests/Helpers/RecordOfSaleSapMessageBuilderTests.cs [hunk 7]
```diff
@@ -313,6 +318,87 @@ public void WhenTransactionTypeIsMaintainHoldings_ThenReturns_SomeFieldsEmptyInS
...
+            && call.GetArgument<IEnumerable<KeyValuePair<string, object>>>(2)!.ToDictionary(c => c.Key, c => c.Value)["{OriginalFormat}"].ToString() == "Creating the record of sale SAP Payload. | _X-Correlation-ID : {_X-Correlation-ID}").MustHaveHappenedOnceExactly();
+
+            A.CallTo(_fakeLogger).Where(call => call.Method.Name == "Log"
+            && call.GetArgument<LogLevel>(0) == LogLevel.Information
+            && call.GetArgument<EventId>(1) == EventIds.CreatedRecordOfSaleSapPayload.ToEventId()
+            && call.GetArgument<IEnumerable<KeyValuePair<string, object>>>(2)!.ToDictionary(c => c.Key, c => c.Value)["{OriginalFormat}"].ToString() == "The record of sale SAP payload created. | _X-Correlation-ID : {_X-Correlation-ID}").MustHaveHappenedOnceExactly();
+        }
+
+        [Test]
+        public void WhenTransactionTypeIsMigrateNewLicence_ThenReturns_SomeFieldsEmptyInSapXmlPayloadCreationTests()
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.EventAggregation.WebJob.Tests/Helpers/RecordOfSaleSapMessageBuilderTests.cs:378]: Test name doesn't look good. Can you update it?

**Review by Vishal14583 — CHANGES_REQUESTED** [review]
Few review comments. Lets have a connect to discuss if required.

File: tests/UKHO.ERPFacade.EventAggregation.WebJob.Tests/Helpers/RecordOfSaleSapMessageBuilderTests.cs (hunk 5 above)
Comment (Dbankar23) [reply tests/UKHO.ERPFacade.EventAggregation.WebJob.Tests/Helpers/RecordOfSaleSapMessageBuilderTests.cs]: Done.

File: tests/UKHO.ERPFacade.EventAggregation.WebJob.Tests/Helpers/RecordOfSaleSapMessageBuilderTests.cs (hunk 7 above)
Comment (sakshi15525) [reply tests/UKHO.ERPFacade.EventAggregation.WebJob.Tests/Helpers/RecordOfSaleSapMessageBuilderTests.cs:378]: We have followed the existing naming convention for the Test Name.

Comment (sakshi15525) [reply tests/UKHO.ERPFacade.EventAggregation.WebJob.Tests/Helpers/RecordOfSaleSapMessageBuilderTests.cs]: Changed to WhenTransactionTypeIsMigrateNewLicence_ThenReturns_SapXmlPayloadWithSomeEmptyFields

**Review by kishorambare — APPROVED** [review]
LGTM

**Review by jrippington — APPROVED** [review]
Looks good

**Pull request merged**